torch/utils/_debug_mode.py

```diff
@@ -668,6 +637,63 @@ def _annotate_fake(tag: str) -> None:
     return None
 
 
+class DebugInterpreter(torch.fx.Interpreter):
+    """
+    Interpreter class for running eager/aot_eager compiled regions when DebugMode is active,
+    instead of using the compiled code. This gives us access to fx.Node metadata to decorate
+    and contextualize DebugMode logs (e.g. nn_module_stack, stack_trace, compiled region boundaries).
+
+    Note: this is currently only enabled with DebugMode(run_compile_with_interpreter=True).
+    """
+
+    def __init__(self, module, backend):
+        super().__init__(module)
+        self.mode = get_active_debug_mode()
+        assert self.mode is not None
+
+        # for tracking initial nn_module_stack
+        self.base_nn_module_stack = list(self.mode.current_nn_module_stack)
+
+        # annotate start of region
+        self.backend = backend
+        self.mode.operators.append(
+            _AnnotateCall(
+                "enter", f"{self.backend} region (compile)", self.mode.call_depth
+            )
+        )
+
+    def run_node(self, n: torch.fx.Node) -> Any:
+        # handling of nn.Module stack
+        if (
+            self.mode.record_nn_module  # type: ignore[attr-defined]
+            and n.op not in ["placeholder", "output"]
+        ):
+            self.mode._handle_fx_nn_module_stack(  # type: ignore[attr-defined]
+                self.base_nn_module_stack,
+                n.meta.get("nn_module_stack", {}),
+                n.meta.get("fwd_nn_module_stack", {}),
+            )
+        return super().run_node(n)
+
+    def run(self, *args, **kwargs):
+        assert self.mode is not None
+        result = super().run(*args)
+
+        # reset nn.Module stack to pre-compiled region value
+        assert len(self.mode.current_nn_module_stack) >= len(self.base_nn_module_stack)
```

Comment:
might be good to emit a warning rather than assert? if there is a bug, users can still see a dump with the rest of the info.